src/rust/engine/graph/src/lib.rs

```diff
@@ -188,6 +188,83 @@ impl<N: Node> InnerGraph<N> {
     None
   }
 
+  ///
+  /// Compute the critical path for this graph.
+  ///
+  /// The critical path is the longest path. For a directed acyclic graph, it is equivalent to a
```

Comment:
`petgraph` has some implementations of shortest path: https://docs.rs/petgraph/0.4.13/petgraph/algo/index.html ... it looks like `astar` might work here?

We use `bellmanford` in other spots: https://github.com/pantsbuild/pants/blob/ef1636d916c1ec97955d487dda3fbd0382802e21/src/rust/engine/graph/src/lib.rs#L168-L189 (this one only needs one path, so could probably just use `astar`?) and https://github.com/pantsbuild/pants/blob/ef1636d916c1ec97955d487dda3fbd0382802e21/src/rust/engine/graph/src/lib.rs#L352-L404 (actually needs multiple paths).

Reply:
I mentioned this in the comment above the implementation:
```
  /// Ideally, we would have used an existing algorithm to calculate the shortest path, such as
  /// petgraph::algo::astar, but this assumes that the weight is on the edges rather than on the
  /// nodes.
  ///
  /// Since in our use of the graph, nodes have a duration as opposed to edges, manually implement
  /// the algorithm.
```

Basically to use astar, I would need to create a new petgraph mapping the vertices to edges, which seemed like it would end up being less clear than just manually coding the thing.

Reply:
@pierrechevalier83 : The second example I linked to supports mapping one graph to another using [filter_map](https://docs.rs/petgraph/0.4.13/petgraph/graph/struct.Graph.html#method.filter_map): it's possible that it doesn't work for this case, but it's pretty flexible?

Reply:
Fair point. It took some fiddling as I had to map the edges to nodes and add a new src and sink nodes, but probably not less clear than previous impl; and it was easy to adapt for the subgraph case. Thanks for the advice 😄